modules/rostests/apitests/user32/CharFuncs.c

```diff
@@ -0,0 +1,597 @@
+     STATUS_ACCESS_VIOLATION},
+    {(PVOID)0xDEADF00D, (PVOID)(0xDEADF00D + 2), NULL,
+     (PVOID)0xDEADF00D, (PVOID)(0xDEADF00D + 2), (PVOID)(0xDEADF00D + 1) /* NULL on Win7 with updates */, NULL,
+     STATUS_ACCESS_VIOLATION},
+    {(PVOID)0xDEADF00D, (PVOID)(0xDEADF00D - 2), (PVOID)(0xDEADF00D - 2),
+     (PVOID)0xDEADF00D, (PVOID)(0xDEADF00D - 2), (PVOID)(0xDEADF00D - 2), (PVOID)(0xDEADF00D - 2),
+     STATUS_SUCCESS},
+    {(PVOID)(0xDEADF00D + 2), (PVOID)0xDEADF00D, (PVOID)0xDEADF00D,
+     (PVOID)(0xDEADF00D + 2), (PVOID)0xDEADF00D, (PVOID)0xDEADF00D, (PVOID)0xDEADF00D,
+     STATUS_SUCCESS},
+};
+
+typedef struct
+{
+    LPCWSTR wszString;
+    LPCWSTR wszResult;
+    LPCSTR szString;
+    LPCSTR szResult;
+    NTSTATUS resStatus;
+} EX_TESTS_CHARNEXT, *PEX_TESTS_CHARNEXT;
+
+EX_TESTS_CHARNEXT TestExceptionCharNext[] =
+{
+    {wszReactOS, wszReactOS + 1,
+      szReactOS,  szReactOS + 1,
+     STATUS_SUCCESS},
+    {NULL, NULL,
+     NULL, NULL,
+     STATUS_ACCESS_VIOLATION},
+    {(PVOID)0xDEADF00D, NULL,
+     (PVOID)0xDEADF00D, NULL,
+     STATUS_ACCESS_VIOLATION},
+
+    {(PVOID)(0xDEADF00D - 2), NULL,
+     (PVOID)(0xDEADF00D - 2), NULL,
+     STATUS_ACCESS_VIOLATION},
+    {(PVOID)(0xDEADF00D + 2), NULL,
+     (PVOID)(0xDEADF00D + 2), NULL,
+     STATUS_ACCESS_VIOLATION},
+};
+
+static LPWSTR AllocStringW(LPWSTR lpszStr, SIZE_T len)
+{
+    LPWSTR str;
+    SIZE_T sz;
+
+    if (!lpszStr)
+        return NULL;
+
+    sz = (len + 1) * sizeof(lpszStr[0]);
+    str = HeapAlloc(GetProcessHeap(), HEAP_ZERO_MEMORY, sz);
+    if (!str)
+    {
+        trace("HeapAlloc failed (error %ld)\n", GetLastError());
+        goto Skip;
+    }
+    StringCbCopyW(str, sz, lpszStr);
+Skip:
+    return str;
+}
+
+static LPSTR AllocStringA(LPWSTR lpszStr, SIZE_T len)
+{
+    LPSTR str;
+    SIZE_T sz, mbs;
+
+    if (!lpszStr)
+        return NULL;
+
+    sz = len + 1;
+    str = HeapAlloc(GetProcessHeap(), HEAP_ZERO_MEMORY, sz);
+    if (!str)
+    {
+        trace("HeapAlloc failed (error %ld)\n", GetLastError());
+        goto Skip;
+    }
+
+    mbs = WideCharToMultiByte(TEST_ACP, 0, lpszStr, -1, NULL, 0, NULL, NULL);
+    if (!mbs || mbs > sz)
+    {
+        HeapFree(GetProcessHeap(), 0, str);
+        str = NULL;
+        trace("WideCharToMultiByte returned %lu (error %ld)\n", mbs, GetLastError());
+        goto Skip;
+    }
+
+    WideCharToMultiByte(TEST_ACP, 0, lpszStr, -1, str, mbs, NULL, NULL);
+Skip:
+    return str;
+}
+
+static void testCharPrev(void)
+{
+    UINT i;
+    SIZE_T len;
+    BOOL b;
+    LPWSTR pchW;
+    LPSTR pchA;
+    INT iRealOffset;
+
+    /* Perform dynamic allocation tests */
+    for (i = 0; i < _countof(TestCharPrev); i++)
+    {
+        PTESTS_CHARPREV pEntry = &TestCharPrev[i];
+        LPWSTR wszStart;
+        LPWSTR wszCurrent;
+        LPSTR szStart;
+        LPSTR szCurrent;
+        LPSTR pchEx;
+
+        len = wcslen(pEntry->lpszStart);
+        wszStart = AllocStringW(pEntry->lpszStart, len);
```

Comment:
This for loop basically contains two "blocks"; that first part with testing 'W', and then the second part with testing 'A'. I'm wondering, whether you could write two inlined helpers for these blocks? so that the for loop becomes, schematically:
```
for (...)
{
    testW(some, params);
    if (pEntry->bWideOnly)
        continue;
    testA(some, params);
}
```
This would help avoiding those DynamicCleanupW/A labels.

Reply:
Good idea, I'll refactor it.